### PR TITLE
[feat] experimental.nn.SyncBatchNorm: initial commit

### PR DESCRIPTION
**Pull request by msbaines**
Fast/simple re-implementation of SyncBatchNorm.

When profiling SSL Vision, I was seeing a majority of cycles spent in
SyncBatchNorm. With this change, I see a 10% to 20% speedup on the
model I was profiling.

When running benchmarks/experimental/sync_batchnorm.py on 8 x V100,
I get a 6x speedup:

<class 'torch.nn.modules.batchnorm.BatchNorm2d'>
Elapsed time is  0.08709120750427246
Elapsed time is  0.12632274627685547
Elapsed time is  0.14095258712768555
Elapsed time is  0.16529417037963867
Elapsed time is  0.1419970989227295
Elapsed time is  0.15166854858398438
Elapsed time is  0.12000870704650879
Elapsed time is  0.17534875869750977
<class 'torch.nn.modules.batchnorm.SyncBatchNorm'>
Elapsed time is  2.5087168216705322
Elapsed time is  2.497001886367798
Elapsed time is  2.5204885005950928
Elapsed time is  2.526789903640747
Elapsed time is  2.5080230236053467
Elapsed time is  2.524489641189575
Elapsed time is  2.513214588165283
Elapsed time is  2.5359973907470703
<class 'fairscale.experimental.nn.sync_batchnorm.SyncBatchNorm'>
Elapsed time is  0.4126114845275879
Elapsed time is  0.39051294326782227
Elapsed time is  0.40685415267944336
Elapsed time is  0.4159870147705078
Elapsed time is  0.42383885383605957
Elapsed time is  0.4080159664154053
Elapsed time is  0.41202712059020996
Elapsed time is  0.42400121688842773

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by min-xu-ai** [comment]
This is very cool! Is the speed up mainly from jit.script or mainly from network traffic optimization?

@prigoyal and I are running into sync_bn issues, but not yet around performance. For us, it is tricky get to it work under activation checkpoint + AMP. There seems to be a corner case bug somewhere. (activation checkpoint matter because of the `track_running_stats`, which we set to false in the no_grad forward pass and reenable back to true in the with_grad forward pass in the checkpoint case.

**Comment by msbaines** [comment]
> This is very cool! Is the speed up mainly from jit.script or mainly from network traffic optimization?
> 
> @prigoyal and I are running into sync_bn issues, but not yet around performance. For us, it is tricky get to it work under activation checkpoint + AMP. There seems to be a corner case bug somewhere. (activation checkpoint matter because of the `track_running_stats`, which we set to false in the no_grad forward pass and reenable back to true in the with_grad forward pass in the checkpoint case.

The PyTorch implementation has a DeviceToHost transfer which this implementation does not. It also uses all_gather while here there are two all_reduces (first could potentially be removed) so the communication over-head is much much less here. The PyTorch implementation uses CUDA kernels so its actually much more efficient in that regard. The jit.script enables some fusing but not so much that its material.

File: benchmarks/experimental/sync_batchnorm.py
Comment (anj-s) [inline benchmarks/experimental/sync_batchnorm.py:33]: Is this a temporary file or do we intend to add this to our benchmark runs on circleCI?

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:70]: s/number of distributed world/devices perhaps?

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:63]: am curious, what is the gain we see by using JIT?

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:66]: any reason we decided to subclass BatchNorm2d? Why not subclass torch.nn.SyncBatchNorm so that we could use the `convert_sync_batchnorm `. Are there things that we have modified in this function?

**Comment by anj-s** [comment]
It also looks like some tests might be failing on older PyTorch versions.

**Comment by prigoyal** [comment]
@msbaines , this is great. I would love to test this out for convergence, speed etc whenever you think makes sense.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:86]: I can see that this might be useful when we have a partial batch. Curious, are there other reasons why we need this allreduce?

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:91]: why don't we just divide by total_count(which should. be the global batch size * (product of all dimensions)? Maybe I don't get how we calculate mean and meansqr but I am not sure I understand why we need to multiple by count here.

File: benchmarks/experimental/sync_batchnorm.py
Comment (msbaines) [reply benchmarks/experimental/sync_batchnorm.py:33]: This file is useful for profiling and was helpful to me when iterating on the code. I'm not yet sure we should run this on circleci. But I plan to use it to use it in the future whenever I make changes.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:63]: I see a reduction in kernel calls in the trace.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:66]: Unfortunately SyncBatchNorm's `convert_sync_batchnorm` explicitly calls `SyncBatchNorm` instead of `cls` so sub-classing it would still require implementing `convert_sync_batchnorm`. I will modify to sub-class `_BatchNorm` although I'm hesitant to sub-class a private class. I didn't want to sub-class SyncBatchNorm because it felt odd to implement `SyncBatchNorm` which sub-classes `SyncBatchNorm`. I was also worried about future changes to `SyncBatchNorm` breaking my sub-class. Felt safer to sub-class on of the other classes. Sub-classes `_BatchNorm` might be the least-worst option.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:86]: Yeah, the primary motivation is to handle partial batches. Another case would be when batch sizes are not the same. This can sometimes happen in NLP due to variable sequence length although we don't use BatchNorm in NLP.

I am contemplating adding a batch-size argument that would avoid this all_reduce and assert if their is a non-matching batch. But this all_reduce does not impact performance that much since it is just a single float and is async. Although, it could matter in a network with high-latency.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:91]: (count / total_count) is this workers contribution to the mean. This is to handle one worker having a larger batch than another. You can think of it like a probability.

So if total_count is 100 and this worker has a count of 40, it will contribute 0.4 towards the aggregate mean.

**Comment by msbaines** [comment]
> @msbaines , this is great. I would love to test this out for convergence, speed etc whenever you think makes sense.

Cool. I would recommend waiting until this change lands just in case we find some issues.

I've done some preliminary testing with @endernewton's code to verify that loss does not change when SyncBatchNorm is replaced with this implementation. Ideally, the results should be close to identical with epsilon differences attributed to the order of operations.

File: benchmarks/experimental/sync_batchnorm.py
Comment (blefaudeux) [reply benchmarks/experimental/sync_batchnorm.py:33]: just FYI (recent discovery of mine), torch has a `torch.utils.benchmark` util which is very nice for these things, removes some boilerplate code and prettifies the outputs. Example [here](https://github.com/fairinternal/xformers/blob/main/benchmarks/benchmark_core.py). I agree that having these small tools is useful, I've not used that enough myself, should have

**Comment by VitaliyLi** [comment]
This is sooo great, @msbaines, 6x speedup!!!!!!

File: tests/experimental/nn/test_sync_batchnorm.py
Comment (anj-s) [inline tests/experimental/nn/test_sync_batchnorm.py:21]: Since this is a fundamental module should we test the GLOO backend as well? if things don't work as expected we can then limit to NCCL support but would help surface any GLOO specific issues.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:62]: I just measured this. Removing the JIT usage increases execution in the benchmark by about 40%.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:90]: Good idea! I considered this. You need to all_reduce proportionally to the local batch_size. I could handle that by all_reduce-ing the sum instead of the mean but then the problem is that you then overflow. I could avoid the overflow by first dividing by a "scale factor" that is order of total_count but was worried about the extra complexity that would add.

**Comment by msbaines** [comment]
While debugging the unit tests, I discovered a bug in PyTorch's computation of the  BatchNorm weight gradient:

https://github.com/pytorch/pytorch/issues/57796

File: tests/experimental/nn/test_sync_batchnorm.py
Comment (msbaines) [reply tests/experimental/nn/test_sync_batchnorm.py:21]: Cool if I handle that in a follow-up change since its marked as experimental? Wanted to unblock testing of the code by others.

Comment (anj-s) [reply tests/experimental/nn/test_sync_batchnorm.py:21]: Absolutely! thanks for following up.

File: fairscale/experimental/nn/sync_batchnorm.py
Comment (anj-s) [reply fairscale/experimental/nn/sync_batchnorm.py]: I think doing the allreduce might be better than requiring an explicit batch size argument.

**Comment by anj-s** [comment]
> While debugging the unit tests, I discovered a bug in PyTorch's computation of the BatchNorm weight gradient:
> 
> [pytorch/pytorch#57796](https://github.com/pytorch/pytorch/issues/57796)

Thank you for the extensive testing! This work has uncovered lot of interesting things.

**Review by anj-s — APPROVED** [review]

**Comment by endernewton** [comment]
Hey, @msbaines nice work on this speed-up! I started trying this on my side on 8x8 GPUs but I could not benchmark the speed on my side as it says "OOM" for the default use case I have (which do not have OOM issue). So I am wondering what is the memory usage of this new op compared to PyTorch's default? If it is less efficient, can it be further optimized in terms of memory usage? Thanks! Attached is the screenshot (I haven't checked the details of the code and just try to use it as an end-user).

image

**Pull request merged**
